tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc

```diff
@@ -398,8 +398,13 @@ Status GrpcServer::Stop() {
       state_ = STOPPED;
       return Status::OK();
     case STARTED:
-      return errors::Unimplemented(
-          "Clean shutdown is not currently implemented");
+      server_->Shutdown();
+      master_service_->Shutdown();
+      worker_service_->Shutdown();
+      eager_service_->Shutdown();
+      state_ = STOPPED;
+      LOG(INFO) << "Server stopped (target: " << target() << ")";
+      return Status::OK();
```

Comment:
I don't think this will lead to a "clean" shutdown. The various services might have requests in flight, and one would need to ensure that they neither leak memory nor hang forever.

Reply:
Do I understand correctly that requests are not accepted since `server_->Shutdown()` call? After this shutdown I stops services. Each service continues processing all his enqueued messages and stops after that. Am I right?

Reply:
Okay, @mrry, do you have any ideas how to make it right? It's really important because without this `stop` functionality we can't properly use server inside other processes.

Reply:
No, I haven't worked on the gRPC part of TensorFlow for many years.

Reply:
I see, @mrry. Anyway, could you please help me a bit with this part? Currently I have the following behavior: when I start and stop server iteratively it fails in ~0.1% cases. It hangs on reading from the queue after shutdown (on `Next()` method).

I've also found out that request enqueueing happens asynchronously and sometimes after shutdown (actually I see that `ENQUEUE_REQUEST` guarded my mutex, but as result of async writing message enqueued after it). Do you know how to fix it? Can we write the message synchronously?

Reply:
Generally speaking, the requirements are very basic. I need to be able to start and stop gRPC servers inside other process without restarting this process.

**TL;DR** The situation is the following. I'd like to add an ability to delegate TensorFlow cluster manager functionality to Apache Ignite cluster manager system (it's java based platform for distributed computing). To do that I basically need to be able to start/stop TensorFlow server inside Apache Ignite Node Process. First of all, we don't have an ability to start gRPC server from Java, but I'm about to add this functionality in #23022. The only one thing stops me from using TensorFlow gRPC server from Java and implement to whole idea is that `stop()` doesn't work so far.

Reply:
Speaking about 'lame duck' approach you described, I think I do exactly these things. I stop gRPC server first of all (`server_->Shutdown()`), then stops master, worker and eager services (`master_service_->Shutdown()`). This leads to enqueueing "SHUTDOWN" alarm message into queue and when this message is read from the queue using `cq_->Next()` method I shutdown the queue. Meanwhile I continue reading trailing messages until `cq_Next()` returns `nullptr` (it works because `Thread` destructor requires the loop with `cq_->Next()` to be completed).

Reply:
I am not familiar with Apache Ignite.  I looked at a little bit of documentation but don't have time to look into it deeply.  We need to be clear about what 'process' means.  Do you want to kill and restart a gRPC server in the same linux process?  Or does Ignite actually kill the linux process and start a new one to bring up the new server Node?  This is an important question because TF has some per-process objects which will persist in the process between server instances, if you really try to restart in the same process.  The main example is memory allocators (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/process_state.h).  This could have strange effects.

Back to your requirements.  I'm guessing that you want Ignite to completely clean up old server instances and create new instances with no prior state.  If that's true, can Ignite not destroy the descendant process in which the old server is running (without waiting for the server to terminate) and create a new one in a new process?

If it's a high enough priority we should be able to make gRPC servers shut down cleanly, but until we try to do it, it's hard to say what kind of problems we might discover.

Reply:
Yes, I'm talking about linux process. I mentioned Apache Ignite just to describe the inspiration of the problem, don't think it has some specific requirements. The idea is to allow user to start a single java application (a java process) and do in this process something like:

```java
while (true) {
  Server server = new Server(...);
  server.start();
  // do something...
  server.stop();
}
```

without memory leaks and hangs. We have the current implementation of that via starting and killing separate Python processes (with `start()` server functionality written in Python) from Java. It works, but seems a bit to tricky compare with an ability to `start` and `stop` servers inside a single java process.

Reply:
Hi, @mrry, @poxvoculi. Do you have any comments regarding this issue?

Reply:
The code in the PR doesn't solve the problem. It would need something like @poxvoculi's suggestion about a lame-duck mode, plus support for cancelling all in-flight requests (which isn't currently implemented). It would also need some refactoring to the device-management code, which uses per-process singletons at various points, so that all the required objects become owned by the server.

The process-based approach that @poxvoculi suggested would be more robust, at the expense of a little more overhead on starting the server, and the absence of a fast path between the parent and child processes. For the latter, it's not clear if that's a requirement.